### PR TITLE
Reaction roles and log categories

### LINKED ISSUE #13: Reaction roles
# Feature
Reaction roles to automate role assignment. Reaction roles will allow regular members to assign select roles to themselves.

## Acceptance Criteria
- [x] Sushi Bot shall allow server owners and server moderators to set up reaction roles, tying a server role to an emoji.
- [x] Sushi Bot shall add a selected emoji to a selected message when a reaction role is set up.
- [x] Users reacting to messages with reaction roles should receive the role tied to the emoji that the user reacted with.
- [x] Users may revoke a role by removing their reaction on a message tied to reaction roles.
- [x] Sushi Bot should allow for different types of reaction roles to be set up. Some of the types should be:
    - [x] regular (users can receive up to all roles attached to a message and revoke a role by removing their reaction),
    - [x] unique (users can only receive one of the roles attached to a message and revoke a role by removing their reaction), and
    - [x] confirm (users can receive a role but not revoke it, only one role per message).

### LINKED ISSUE #32: Log categories
# Feature Request
Categories for the log to be sorted into, such as messages (edited, deleted) and people joining/leaving the server.

## Acceptance Criteria
- [x] Sushi bot shall have 3 different categories, and they should be named:
  - [x] Messages
  - [x] Members
  - [x] Profiles
- [x] Sushi bot shall let server administrators opt in or out of different categories.

## Additional Context
For the sake of not flooding the log chat with unimportant log messages and to make it easier for server administrators to keep track of what's happening while they're gone.

### PR DESCRIPTION
**Pull request by MirageAegis**
# What Does This PR Do?
This PR brings reaction roles to Sushi Bot. It also refactors the schema wrapper classes used when interfacing with MongoDB. The PR also enhances moderation logs by adding categories that server administrators can opt in- or out of.

## Acceptance Criteria Met
- Sushi Bot has a command that lets server administrators set up reaction roles.
- Sushi Bot adds the selected emojis as buttons when a reaction roles message is set up.
- Users receive the role tied to an emoji when interacting with a button with that emoji.
- Users can revoke a reaction role by interacting with a button that has an emoji tied to a role that they have.
- Sushi Bot allow server administrators to set up different types of reaction roles.
  - Regular reaction roles allow users to choose multiple roles.
  - Unique reaction roles allow users to choose only one of the roles.
  - Confirm reaction roles disallows users from revoking the role.

- Sushi Bot has three moderation log categories, those being:
  - members,
  - messages, and
  - profiles.
- Sushi Bot lets server administrators configure all three moderation logs separately.

## Related Issues
Closes #13 
Closes #32

### CODE REVIEW

**Comment by MirageAegis** [comment]
Update: #32 will be implemented with this PR as well

File: src/schemas/server.ts [hunk 1]
```diff
@@ -22,11 +22,16 @@
...
+import { Schema, HydratedDocument, Model, model } from "mongoose";
 
 const serverSchema: Schema = new Schema({
     _id: String,
-    logs: String,
+    logs: {
+        members: String,
+        messages: String,
+        profiles: String
+    },
```
Comment (MirageAegis) [inline src/schemas/server.ts:34]: First of all, the server schema has been changed to accommodate moderation logs categories...

File: src/schemas/server.ts [hunk 2]
```diff
@@ -35,9 +40,33 @@ const serverSchema: Schema = new Schema({
...
+    reactionMessages: {
+        type: Map,
+        of: {
+            channel: String,
+            mode: String,
+            reactionRoles: {
+                type: Map,
+                of: String
+            }
+        },
```
Comment (MirageAegis) [inline src/schemas/server.ts:53]: And reaction roles.

File: src/schemas/server.ts [hunk 3]
```diff
@@ -177,37 +245,156 @@ export class Server {
...
+        return this.data.logs?.profiles;
+    }
+
+    public set profileLogs(id: Snowflake) {
+        this.data.logs = {
+            members: this.data.logs.members,
+            messages: this.data.logs.messages,
+            profiles: id
+        };
     }
```
Comment (MirageAegis) [inline src/schemas/server.ts:305]: Setters and getters for the logs configurations have been updated.

File: src/schemas/server.ts [hunk 4]
```diff
@@ -177,37 +245,156 @@ export class Server {
...
+            mode: mode,
+            reactionRoles: reactionRoles
+        };
+
+        messages.set(message, roles);
+    }
+
+    public deleteReactionRoles(message: Snowflake): void {
+        this.data.reactionMessages.delete(message);
     }
```
Comment (MirageAegis) [inline src/schemas/server.ts:399]: The reaction roles have their own setters and getters along with a getter that retrieves all reaction roles, and a delete method for removing reaction roles.

File: src/util/error-handler.ts [hunk 5]
```diff
@@ -99,4 +100,56 @@ export const defaultErrorHandler: ErrorHandler = async (ctx: ChatInputCommandInt
...
+                           `Server: ${ctx.guild.name}\n` +
+                           `User: ${user.id} (${user.username})\n\n` +
+                           // Error name, and error code if it's a Discord API error
+                           `${err.name}${err instanceof DiscordAPIError ? `: ${err.code} (${err.message})` : ""}\n` +
+                           "```";
+    
+    await getAdminLogsChannel().send(report);
+    
+    // ----- !ERROR LOG -----
 };
```
Comment (MirageAegis) [inline src/util/error-handler.ts:155]: The reaction roles have their own basic error handler.

File: src/commands/configuration/config-subcommands/reactionroles.ts [hunk 6]
```diff
@@ -0,0 +1,862 @@
...
+const MIN_RR: number = 1;
+const MAX_RR: number = 15;
+
+const name: string = "reactionroles";
+
+export const command: Subcommand = {
+    // Command headers
+    data: new SlashCommandSubcommandBuilder()
+        .setName(name)
+        .setDescription("Starts a reaction roles setup wizard"),
```
Comment (MirageAegis) [inline src/commands/configuration/config-subcommands/reactionroles.ts:66]: The `/config reactionroles` command is a lengthy one that walks users through the process of creating reaction roles.

It consists of seven steps:
1. Choosing a reaction role style.
2. Choosing a message format (plain text or embedded message).
3. Creating the message.
4. Creating reaction roles.
5. Reviewing the message and reaction roles.
6. Sending the reaction roles message.
7. Saving the data to the database.

File: src/commands/configuration/config-subcommands/memberlogs.ts [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+export const command: Subcommand = {
+    // Command headers
+    data: new SlashCommandSubcommandBuilder()
+        .setName(name)
+        .setDescription("Sets or unsets a channel for member logging")
+        .addChannelOption(o => 
+            o.setName("channel")
+                .setDescription("The channel to use as a member logs channel")
+                .addChannelTypes(ChannelType.GuildText)
+        ),
```
Comment (MirageAegis) [inline src/commands/configuration/config-subcommands/memberlogs.ts:48]: The `/config logs` command has been copied into `/config memberlogs`, `/config messagelogs`, and `/config profilelogs`; each configuring logs for their own category.

File: src/events/reactionroles.ts [hunk 8]
```diff
@@ -0,0 +1,210 @@
...
+    } else {
+        // The confirm style only allows users to get a role
+        await member.roles.add(role);
+
+            await button.reply({
+                content: `Added <@&${role}>`,
+                ephemeral: true
+            });
+    }
+};
```
Comment (MirageAegis) [inline src/events/reactionroles.ts:136]: This is the event handler for reaction roles.

File: src/events/reactionroles.ts [hunk 9]
```diff
@@ -0,0 +1,210 @@
...
+    // Otherwise initiate
+    refreshInitiated = true;
+
+    await refreshReactionRoles(client);
+
+    // Create a co-routine that triggers every six hours
+    setInterval(async (): Promise<void> => {
+        await refreshReactionRoles(client);
+    }, SIX_HOURS);
+};
```
Comment (MirageAegis) [inline src/events/reactionroles.ts:210]: The stored reaction roles are set to refresh every six hours, to remove deleted reaction roles.

File: example.env [hunk 10]
```diff
@@ -0,0 +1,15 @@
...
+USER_REPORTS_CHANNEL_ID={THE ID OF THE CHANNEL TO OUTPUT USER REPORTS}
+LOGS_CHANNEL_ID={THE ID OF THE CHANNEL TO OUTPUT ERROR LOGS}
+VTUBER_ROLE_ID={THE ID OF THE VTUBER ROLE}
+
+# MongoDB related
+MONGO_DB_URI={THE MONGODB ACCESS URI}
+BLACKLIST_ID={THE MONGO ID FOR THE BLACKLIST}
+
+# Tenor related
+TENOR_API_KEY={THE TENOR API KEY FOR GIFS}
```
Comment (MirageAegis) [inline example.env:15]: Lastly, an example `.env` file has been added.

**Review by MirageAegis — COMMENTED** [review]
This review will walk through all noteworthy changes this PR brings.

**Review by zahatikoff — APPROVED** [review]

**Pull request merged**